### PR TITLE
doc: initial draft of RBD mirroring admin documentation

### PR DESCRIPTION
**Pull request by dillaman**
Fixes: #15041

Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

**Comment by dillaman** [comment]
@bniver in case you wish to lend an extra set of eyes ...

**Comment by dillaman** [comment]
http://docs.ceph.com/docs/wip-15041/rbd/rbd-mirroring/

File: doc/rbd/rbd-mirroring.rst [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+RBD images can be asynchronously mirrored between two Ceph clusters.  This
+capability uses the RBD journaling image feature to ensure crash-consistent
+replication between clusters.  Mirroring is configured on a per-pool basis
+within peer clusters and can be configured to automatically mirror all pool
+images within a pool or only a specific subset of images.  Mirroring is
+configured using the ``rbd`` command.  The ``rbd-mirror`` daemon is responsible
+for pulling image updates from the remote, peer cluster and applying them to
+the image within the local cluster.
+
+.. important:: To use use RBD mirroring, you must have two Ceph clusters and
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: s/use use/use/

File: doc/rbd/rbd-mirroring.rst [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+   the ability for the ``rbd-mirror`` daemon to connect to both clusters
+   simultaneously.
+
+Pool Configuration
+==================
+
+The following procedures demonstrate how perform the basic administrative
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: s/how perform/how to perform/

File: doc/rbd/rbd-mirroring.rst [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+Disable Mirroring
+-----------------
+
+To disable mirroring on a pool with ``rbd``, specify the ``mirror pool disable``
+command and the pool name::
+
+        rbd mirror pool disable {pool-name}
+
+When mirroring is disabled on the pool, any images where mirroring was
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: s/where/for which/ ?

File: doc/rbd/rbd-mirroring.rst [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+explicitly enabled will also be disabled.
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: When mirroring is disabled on a pool in this way, mirroring will also be disabled on any images (within the pool) for which mirroring was enabled explicitly.

File: doc/rbd/rbd-mirroring.rst [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+        rbd mirror pool peer add {pool-name} {client-name}@{cluster-name}
+
+For example::
+
+        rbd --cluster local mirror pool peer add image-pool client.remote@remote
+        rbd --cluster remote mirror pool peer add image-pool client.local@local
+
+.. note:: The peer cluster name corresponds to a Ceph configuration file of the
+   same name (e.g. /etc/ceph/remote.conf)
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: This seems to imply that `/etc/ceph/remote.conf` needs to be present on the host where we are running the `rbd` command. But ordinarily it would only be present on the remote cluster itself, right?

Comment (dillaman) [inline doc/rbd/rbd-mirroring.rst]: That's why I included the "These procedures assume two clusters, named "local" and "remote", are reachable from a single host for clarity." bit.  However, as noted, the rbd-mirror daemon must have the ability to connect to both clusters.

File: doc/rbd/rbd-mirroring.rst [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+For example::
+
+        rbd --cluster local mirror pool peer remove image-pool 55672766-c02b-4729-8567-f13a66893445
+        rbd --cluster remote mirror pool peer remove image-pool 60c0e299-b38f-4234-91f6-eed0a367be08
+
+Image Configuration
+===================
+
+Unlike pool configuration, image configuration only needs to be performed against
+a single Ceph cluster (i.e. the primary data center Ceph cluster).
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: The parenthetical implies that the image configuration commands *must* be run against the primary cluster, but up until this point we have been discussing "peer" clusters - not primary and secondary.

Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: If the image configuration commands can be run against either of the clusters, the parenthetical could be dropped.

Comment (dillaman) [inline doc/rbd/rbd-mirroring.rst]: I can add something to clarify.   Primary/secondary is a per-image concept.   For example, enabling mirroring on an image implicitly promotes it to primary.  However, that image could have been promoted on the local or remote cluster.

File: doc/rbd/rbd-mirroring.rst [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+Image Promotion and Demotion
+----------------------------
+
+An image can only be primary within a single Ceph cluster.  Attempting to
+access a non-primary image will result in an error.  In a failover scenario,
+access to the primary image should be stopped (e.g. power down the VM or remove
+the associated drive from a VM), demote the current primary image, promote
+the new primary image, and resume access to the image on the alternate
+cluster.
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: This paragraph is confusing because it suddenly introduces the concept of "primary" and "non-primary" images. Also, the first sentence "An image can only be primary within a single Ceph cluster" doesn't make sense to me.
It would be great if you could clarify the primary/non-primary dichotomy for images as well as for clusters.

Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: The sentence starting with "In a failover scenario, . . ." makes it sound like failover is manual - correct? In order to failover to the mirror, the Ceph admin has to perform the actions listed (stop access, demote, promote, resume) manually?

File: doc/rbd/rbd-mirroring.rst [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+For example::
+
+        rbd --cluster remote mirror image promote image-pool/image-1
+
+.. tip:: Since the primary / secondary status is per-image, it is possible to
+   have two clusters split the IO load and stage failover / failback.
+.. note:: Promotion can be forced using the ``--force`` option.  This will result
+   in a split-brain scenario between the two peers and the image will no longer
+   be in-sync until a `force resync command`_ is issued.
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: Here I'm missing an example of a situation where inducing split-brain via `--force` would be useful.

Comment (dillaman) [inline doc/rbd/rbd-mirroring.rst]: Example: cluster A died -- since I cannot demote the image, I need to force promote the image in the other cluster.

File: doc/rbd/rbd-mirroring.rst (hunk 7 above)
Comment (dillaman) [inline doc/rbd/rbd-mirroring.rst]: RBD cannot coordinate failover -- way too many moving parts outside our control.  For example, Cinder failover support in OpenStack will coordinate all of this (eventually) by disconnecting the drive, demoting, promoting, and reconnecting the drive.

File: doc/rbd/rbd-mirroring.rst [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+replication between clusters.  Mirroring is configured on a per-pool basis
+within peer clusters and can be configured to automatically mirror all pool
+images within a pool or only a specific subset of images.  Mirroring is
+configured using the ``rbd`` command.  The ``rbd-mirror`` daemon is responsible
+for pulling image updates from the remote, peer cluster and applying them to
+the image within the local cluster.
+
+.. important:: To use use RBD mirroring, you must have two Ceph clusters and
+   the ability for the ``rbd-mirror`` daemon to connect to both clusters
+   simultaneously.
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: This makes it sound like there is only a single `rbd-mirror` daemon for both clusters. After reading the document I'm confused about where this daemon actually runs. I assume a single instance of the daemon runs on each of the peer clusters, but I'm not sure.

File: doc/rbd/rbd-mirroring.rst [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+   the resync asynchronously.
+
+rbd-mirror Daemon
+=================
+
+The ``rbd-mirror`` daemon is responsible for watching image journals on a
+remote, peer cluster and replaying the journal events against the local
+cluster.  The journal events recorded by the RBD image journaling feature
+include all modifications to the image in-order to ensure that a
+crash-consistent mirror of the remote image is available locally
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: s/locally/locally./ but I suggest rewording the last sentence like so:

The RBD image journaling feature records all modifications to the image in the order they occur. This ensures that a crash-consistent mirror of the remote image is available locally.

File: doc/rbd/rbd-mirroring.rst (hunk 8 above)
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: OK, so add a parenthetical to the end of the note like this?

(Force promotion is needed when one of the peer clusters dies.)

File: doc/rbd/rbd-mirroring.rst (hunk 6 above)
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: If primary/secondary is per-image, not per-cluster, then I would not recommend writing "the primary data center Ceph cluster" . . .

File: doc/rbd/rbd-mirroring.rst (hunk 7 above)
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: OK, we can assume this will be obvious to the document's intended audience.

File: doc/rbd/rbd-mirroring.rst (hunk 5 above)
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: Ah! I understood "reachable" to mean "pingable".

**Comment by dillaman** [comment]
@smithfarm updated draft pushed

File: doc/rbd/rbd-mirroring.rst [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+``rbd`` command).
+
+Enable Image Journaling Support
+-------------------------------
+
+RBD mirroring uses the RBD journaling feature to ensure that the replicated
+image always remains crash-consistent.  Before an image can be mirrored to
+a peer cluster, the journaling feature must be enabled.  The feature can be
+enabled at image creation time by including
+``--image-feature exclusive-lock,journaling`` option to the ``rbd`` command.
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: s/by including/by providing the/

File: doc/rbd/rbd-mirroring.rst [hunk 12]
```diff
@@ -0,0 +1,250 @@
+===============
+ RBD Mirroring
+===============
+
+.. index:: Ceph Block Device; mirroring
+
+RBD images can be asynchronously mirrored between two Ceph clusters.  This
+capability uses the RBD journaling image feature to ensure crash-consistent
+replication between clusters.  Mirroring is configured on a per-pool basis
+within peer clusters and can be configured to automatically mirror all pool
```
Comment (jdurgin) [inline doc/rbd/rbd-mirroring.rst]: s/all pool/all/

File: doc/rbd/rbd-mirroring.rst [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+images within a pool or only a specific subset of images.  Mirroring is
+configured using the ``rbd`` command.  The ``rbd-mirror`` daemon is responsible
+for pulling image updates from the remote, peer cluster and applying them to
+the image within the local cluster.
+
+.. important:: To use RBD mirroring, you must have two Ceph clusters, each
+   running the ``rbd-mirror`` daemon.
```
Comment (jdurgin) [inline doc/rbd/rbd-mirroring.rst]: Perhaps note that this requires jewel or later here too

File: doc/rbd/rbd-mirroring.rst [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+        rbd mirror image promote {pool-name}/{image-name}
+
+For example::
+
+        rbd --cluster remote mirror image promote image-pool/image-1
+
+.. tip:: Since the primary / non-primary status is per-image, it is possible to
+   have two clusters split the IO load and stage failover / failback.
+
+.. note:: Promotion can be forced using the ``--force`` option.  Forced
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: only one space after period

File: doc/rbd/rbd-mirroring.rst [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+   promotion is needed when the demotion cannot be propagated to the peer
+   Ceph cluster (e.g. Ceph cluster failure, communication outage).  This will
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: dtto

Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: In high school I was taught to always write two spaces after a period, but nowadays one is standard: http://theworldsgreatestbook.com/how-many-spaces-after-a-period/

Please fix throughout the document.

File: doc/rbd/rbd-mirroring.rst [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+
+The pool configuration steps should be performed on both peer clusters. These
+procedures assume two clusters, named "local" and "remote", are accessible from
+a single host for clarity.
+
+See the `rbd`_ manpage for additional details of how to connect to different
+Ceph clusters.
+
+.. note:: The cluster name in the following examples corresponds to a Ceph
+   configuration file of the same name (e.g. /etc/ceph/remote.conf).
```
Comment (jdurgin) [inline doc/rbd/rbd-mirroring.rst]: maybe point out keyring name too, and reference http://docs.ceph.com/docs/master/rados/configuration/ceph-conf/#running-multiple-clusters or http://docs.ceph.com/docs/master/rados/configuration/auth-config-ref/#keys

File: doc/rbd/rbd-mirroring.rst (hunk 15 above)
Comment (dillaman) [inline doc/rbd/rbd-mirroring.rst]: Ditto -- I thought you were implying that two spaces vs one was some magic with RST.

File: doc/rbd/rbd-mirroring.rst (hunk 16 above)
Comment (dillaman) [inline doc/rbd/rbd-mirroring.rst]: Maybe just point to the `running-multiple-clusters` since otherwise I would probably also have to document specifying the client name / id.  I think it starts distracting from core message.

File: doc/rbd/rbd-mirroring.rst [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+.. note:: The ``rbd`` command only flags the image as requiring a resync. The
+   local cluster's ``rbd-mirror`` daemon process is responsible for performing
+   the resync asynchronously.
+
+rbd-mirror Daemon
+=================
+
+The ``rbd-mirror`` daemon is responsible for watching image journals on a
+remote, peer cluster and replaying the journal events against the local
+cluster.  The RBD image journaling feature records all modifications to the
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: I know there are two, but the phrase "The rbd-mirror daemon" causes the reader to think in terms of a single daemon. How about this:

The two `rbd-mirror` daemons (one on each peer cluster) are responsible for watching image journals on the remote, peer cluster and . . .

File: doc/rbd/rbd-mirroring.rst [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+image in the order they occur. This ensures that a crash-consistent mirror of
+the remote image is available locally.
+
+The ``rbd-mirror`` daemon is available within the optional ``rbd-mirror``
+distribution package.
+
+.. important:: The ``rbd-mirror`` daemon requires the ability to connect
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: Each `rbd-mirror` daemon requires . . .

File: doc/rbd/rbd-mirroring.rst [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+   to both clusters simultaneously.
+.. warning:: Only run a single ``rbd-mirror`` daemon per Ceph cluster. A
+   future Ceph release will add support for horizonatal scale-out of the
```
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: s/horizonatal/horizontal/

File: doc/rbd/rbd-mirroring.rst (hunk 15 above)
Comment (smithfarm) [inline doc/rbd/rbd-mirroring.rst]: Nope. I'm a pedant.

**Comment by smithfarm** [comment]
@dillaman Thanks for taking my comments under consideration! Upon consideration of the remaining nits: `Reviewed-by: Nathan Cutler <ncutler@suse.com>`

**Comment by dillaman** [comment]
@smithfarm much appreciated for your review

File: doc/rbd/rbd-mirroring.rst [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+The pool configuration steps should be performed on both peer clusters. These
+procedures assume two clusters, named "local" and "remote", are accessible from
+a single host for clarity.
+
+See the `rbd`_ manpage for additional details of how to connect to different
+Ceph clusters.
+
+.. note:: The cluster name in the following examples corresponds to a Ceph
+   configuration file of the same name (e.g. /etc/ceph/remote.conf).  See the
+   `ceph-conf`_ documentation for how configure multiple clusters.
```
Comment (jdurgin) [inline doc/rbd/rbd-mirroring.rst]: s/how/how to/
